### ISSUE TITLE
docs(core): fix stale training entry point and broken examples

### ISSUE THREAD

**Issue by CedrickArmel** [CedrickArmel/radiologist#238]
Part of the docs accuracy epic. **Content only** — do not change any Python, config, or CLI code in this issue.

Target file: `radiologist-core/README.md`

`docs/pkg/core.md` is a pure `{% include-markdown "../../radiologist-core/README.md" %}` shim, so editing the README is the whole job — no change to `docs/pkg/core.md` is needed.

---

### 1. The quick start invokes an entry point that does not exist (lines 57–80)

```bash
cd radiologist-core
uv run --active python -m radiologist.core.train
```

`radiologist-core/src/radiologist/core/train.py` has **no** `if __name__ == "__main__"` block and **no** `@hydra.main` decorator — it exports only the `train(cfg)` function. `radiologist-core/pyproject.toml` has no `[project.scripts]` table.

The real Hydra entry point is `radiologist.cli.groups.core:train_main` at `radiologist-cli/src/radiologist/cli/groups/core.py:101`:

```python
@hydra.main(
    config_path="pkg://radiologist.core.configs",
    config_name="train",
    version_base="1.3",
)
def train_main(cfg: DictConfig) -> Optional[float]:
```

Lines 66–72 and 76–80 repeat the same wrong invocation.

### 2. The `LModule` example does not run (lines 88–101)

The README shows:

```python
module = LModule(
    net=resnet50(num_classes=3),
    loss=FocalLoss(gamma=2, alpha=1),
    metric=MulticlassFBetaScore(...),
    optimizer=partial(AdamW, ...),
    scheduler=partial(sequential_scheduler, ...),
    trainable_layers=None,
    priors=None,
)
```

The real signature is `radiologist-core/src/radiologist/core/module.py:61`:

```python
def __init__(self, cfg: DictConfig) -> None:
```

**One** argument. Every collaborator is built internally via `hydra.utils.instantiate(cfg.get(...))` at `module.py:66–74`. The documented call raises `TypeError`. `train.py` itself does `LModule(cfg=cfg.get("module"))`.

### 3. The `WebDatasetDataModule` example does not run (lines 114–123)

The README passes four arguments. `data/datamodule.py`'s `__init__` has **eight** required parameters with no defaults: `shard_root`, `split_manifest_uri`, `label_map`, `train_transform`, `eval_transform`, `train_loader`, `eval_loader`, `batch_size`. The snippet raises `TypeError`.

### 4. Two wrong rows in the `## Configuration reference` table

- **Line 197**: `` | `module/resnet50.yaml` | ResNet-50 backbone, `num_classes` interpolated from datamodule | `` — inverted. `configs/module/resnet50.yaml` **hardcodes** `num_classes: 3`; it is `configs/module/metric/fbeta_score.yaml` that interpolates `${module.net.num_classes}`.
- **Line 202**: the `callbacks/default.yaml` cell lists five callbacks. The actual defaults list in `configs/callbacks/default.yaml` has six — `onnx_export` is missing from the doc.

### 5. Stale registry CLI name (line 185)

> "or the `radiologist-registry push` / `promote` CLI"

That console script no longer exists; `radiologist-registry/pyproject.toml` has no `[project.scripts]`.

### 6. Stale package-layout comment (line 35)

`` train.py  # Hydra entry point and train() function `` — `train.py` holds only `train()`.

---

### AC1 — Replace lines 57–80 (`### Quick start`, `### Evaluation only`)

````markdown
### Quick start

This package ships no CLI of its own. Training is the `core` group of the
unified `radiologist` command, which lives in `radiologist-cli`:

```bash
pip install radiologist-cli    # the core group is a hard dependency, no extra needed
radiologist core
```

Defaults are composed from `src/radiologist/core/configs/train.yaml`. The group
takes arbitrary Hydra overrides — everything after `core` is forwarded to Hydra
verbatim:

```bash
radiologist core \
    trainer.max_epochs=30 \
    datamodule.shard_root=gs://my-bucket/shards/ \
    datamodule.split_manifest_uri=gs://my-bucket/manifests/manifest-abc123.jsonl \
    tags=[baseline] \
    seed=42
```

`tags` is not optional in practice: `extras.enforce_tags: true` fails the run
fast when it is unset, so W&B runs stay searchable.

On success the command emits a six-key record (`run_id`, `best_ckpt_path`,
`det_onnx_path`, `mcd_onnx_path`, `det_qualified_name`, `mcd_qualified_name`);
add the global `--output json` flag for machine-readable output.

### Evaluation only

`eval.yaml` is a second top-level config, selected with Hydra's `--config-name`
flag rather than a separate subcommand:

```bash
radiologist core --config-name eval ckpt_path=/path/to/checkpoint.ckpt
```

It sets `train: false`, `test: true`, `loggers: null`, and makes `ckpt_path`
mandatory (`???`).
````

### AC2 — Replace lines 86–101 (the `### \`LModule\`` intro + example)

Keep the existing prose at lines 103–108 (`trainable_layers` / priors behavior) unchanged — it is accurate.

````markdown
A backbone-agnostic `LightningModule`. It takes a **single** `DictConfig` and
instantiates every collaborator itself through `hydra.utils.instantiate`, so the
same object is constructible from the packaged config tree or from a config you
build in Python:

```python
from omegaconf import OmegaConf
from radiologist.core import LModule

module_cfg = OmegaConf.create(
    {
        "net": {"_target_": "torchvision.models.resnet50", "num_classes": 3},
        "loss": {
            "_target_": "radiologist.core.FocalLoss",
            "gamma": 2, "alpha": 1, "to_onehot_y": True,
            "use_softmax": True, "reduction": "mean",
        },
        # metric/optimizer/scheduler are *partial* factories — `_partial_: true`
        "metric": {
            "_target_": "torchmetrics.classification.MulticlassFBetaScore",
            "_partial_": True, "num_classes": 3, "beta": 1.0, "average": "macro",
        },
        "optimizer": {"_target_": "torch.optim.AdamW", "_partial_": True, "lr": 1e-3},
        "scheduler": None,          # None is allowed: no LR schedule
        "trainable_layers": None,   # None = full re-init; dict of dot-paths = fine-tune
        "priors": None,             # None = taken from the datamodule at setup time
    }
)
lm = LModule(cfg=module_cfg)
```

The config contract `LModule` enforces:

| Key | Requirement |
|---|---|
| `net` | fully instantiated `nn.Module`; called as `net(x)` and expected to return logits |
| `loss` | fully instantiated callable; called as `loss(logits, target)` where `target` is a `torch.long` class-index tensor of shape `(B,)` |
| `metric` | **partial** factory (`_partial_: true`). Called with no arguments twice (val + test) and must return a `torchmetrics.Metric` exposing `compute_on_cpu`, `sync_on_compute`, and `process_group` — those three are copied onto the internal `MeanMetric` loss trackers |
| `optimizer` | **partial** factory; called as `optimizer(params=...)` |
| `scheduler` | **partial** factory or `None`; called as `scheduler(optimizer=...)`, stepped per **step**, not per epoch |
| `trainable_layers` | `None`, or a mapping of dot-path → `None` (unfreeze whole submodule) or list of ints (unfreeze those indices) |

Note the metric is updated with **hard predicted class indices**
(`logits.argmax(dim=1)`), not probabilities — probability-based metrics such as
`MulticlassAUROC` will not behave meaningfully here.
````

### AC3 — Replace lines 114–123 (the `WebDatasetDataModule` example)

Keep lines 125–127 (the `label_map` / dataloader-pipeline prose) unchanged.

````markdown
Streams images from WebDataset tar shards. Handles class-balanced sampling
automatically. Eight arguments are required — the two transform pipelines and
the two `WebLoader` partial factories have no defaults, so the practical way to
build one outside the CLI is to instantiate it from the packaged config:

```python
from hydra import compose, initialize_config_module
from hydra.utils import instantiate

with initialize_config_module("radiologist.core.configs", version_base="1.3"):
    cfg = compose(
        config_name="train",
        overrides=[
            "datamodule.shard_root=gs://bucket/shards/",
            "datamodule.split_manifest_uri=gs://bucket/manifests/manifest-abc123.jsonl",
            "datamodule.batch_size=32",
        ],
    )

dm = instantiate(cfg.datamodule)
```

Each batch is a dict with keys `input` (image tensor), `target` (`torch.long`
class index), and `key` (the WebDataset sample key).

The `label_map` collapses raw ETL folder names (e.g. `normal`, `covid`) into
model class names (e.g. `healthy`, `viral`). This decouples the dataset's folder
structure from the model's output space.
````

### AC4 — New `## Using the public API`

Insert **after** `## Exporting and promoting a trained model` (after line 185), **before** `## Configuration reference`.

````markdown
## Using the public API

Everything below is reachable from `radiologist.core`'s `__init__.py` — the
package's supported surface. Nothing here needs the CLI or a Hydra run
directory; use it when you are embedding training in a notebook, a sweep
driver, or another service.

```python
from radiologist.core import (
    AttributionCallback,
    BestMetricCallback,
    FocalLoss,
    LModule,
    OnnxExportCallback,
    train,
    WandbDefineSummaryCallback,
    WebDatasetDataModule,
)
```

`radiologist.core.registry` exports one further symbol, `export_onnx`, from its
own `__init__.py`.

### Driving a full run from Python

`train(cfg)` is the whole training/evaluation orchestration — seeding, callback
and logger instantiation, datamodule and module construction, `fit`, `test` —
and it is a plain function over a `DictConfig`. Compose the config yourself
instead of going through `@hydra.main`:

```python
from hydra import compose, initialize_config_module
from radiologist.core import train

with initialize_config_module("radiologist.core.configs", version_base="1.3"):
    cfg = compose(
        config_name="train",
        overrides=[
            "trainer.max_epochs=5",
            "tags=[notebook]",
            "datamodule.shard_root=gs://bucket/shards/",
        ],
    )

metric_dict, object_dict = train(cfg)
print(metric_dict["val_score"])
print(object_dict["trainer"].checkpoint_callback.best_model_path)
```

`train` returns `(metric_dict, object_dict)`: the first is the merged
train/test callback metrics, the second holds the live `trainer`, `module`,
`datamodule`, `callbacks`, and `loggers` objects. `paths.output_dir` resolves
from Hydra's runtime output dir, so under `compose()` (no run directory) point
Lightning somewhere explicit with `trainer.default_root_dir=...` if you need
artifacts on disk.

### Composing the pieces by hand

The building blocks are independent — you can wire them into your own Lightning
`Trainer` without `train()` at all:

```python
import lightning as L
from radiologist.core import BestMetricCallback, LModule

lm = LModule(cfg=module_cfg)            # see the LModule contract above
trainer = L.Trainer(
    max_epochs=5,
    accelerator="cpu",
    callbacks=[BestMetricCallback(monitor="val_score", mode="max")],
)
trainer.fit(lm, datamodule=dm)
```

### Checkpoint → ONNX, without W&B

`export_onnx` is pure local file work; it performs no registry calls:

```python
from radiologist.core.registry import export_onnx

result = export_onnx(
    ckpt_path="/path/to/checkpoint.ckpt",
    run_id="wandb-run-id",
    input_shape=(1, 3, 224, 224),
    classes=["healthy", "viral", "opacity"],
    cam_target_layer="layer4.1.conv2",
    out_dir="/tmp/onnx-export",
    opset=18,
)
result.det_path, result.mcd_path
```

Uploading and promoting those files is `radiologist-registry`'s job — see
[Using the public API](../radiologist-registry/README.md#using-the-public-api)
there.

### Loss and callbacks standalone

`FocalLoss` is an ordinary `nn.Module` and the four callbacks are ordinary
Lightning callbacks; both are usable in any training loop:

```python
from radiologist.core import FocalLoss

loss = FocalLoss(gamma=2.0, alpha=1.0, use_softmax=True, reduction="mean")
```
````

### AC5 — New `## Extending via Hydra`

Insert **after** `## Configuration reference` (after line 202), **before** `## Optional extras`.

`````markdown
## Extending via Hydra

You do not need to fork or subclass anything to swap the loss, the metric, or
the whole backbone. Every one of those is a Hydra **config group** whose members
are selected by name and instantiated from a `_target_`, so your own class in
your own package is a first-class option.

The packaged configs live inside the installed wheel (`config_path` is
`pkg://radiologist.core.configs`), so you add options by pointing Hydra at an
extra search path. Two equivalent ways:

```bash
# as a config override (works anywhere, including inside a sweep)
radiologist core hydra.searchpath=[file:///abs/path/to/myconfigs] module/loss=my_loss

# as a Hydra CLI flag
radiologist core --config-dir /abs/path/to/myconfigs module/loss=my_loss
```

Your directory mirrors the packaged tree:

```
myconfigs/
├── module/
│   ├── my_net.yaml           # a whole backbone (top-level `module` group)
│   ├── loss/my_loss.yaml     # a loss   (`module/loss` group)
│   └── metric/my_metric.yaml # a metric (`module/metric` group)
```

> **The one thing that will silently bite you.** Every member of the `module`
> group and its sub-groups must start with a `# @package module` header, exactly
> as the shipped configs do. Hydra otherwise infers the package from the
> directory name and your `loss:` key lands at `module.loss.loss`, where
> `LModule` never looks — the run then fails with a confusing instantiation
> error instead of a config error. The shipped `callbacks/*.yaml` use
> `# @package callbacks`, and `datamodule/default.yaml` / `trainer.yaml` use
> `# @package _global_`; match whichever group you are extending.

### A custom loss

```yaml
# myconfigs/module/loss/my_loss.yaml
# @package module
loss:
  _target_: my_package.losses.LabelSmoothedCE
  smoothing: 0.05
```

```python
# my_package/losses.py
import torch
from torch import nn


class LabelSmoothedCE(nn.Module):
    def __init__(self, smoothing: float = 0.0) -> None:
        super().__init__()
        self._inner = nn.CrossEntropyLoss(label_smoothing=smoothing)

    def forward(self, logits: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
        # logits: (B, num_classes) float; target: (B,) int64 class indices
        return self._inner(logits, target)
```

```bash
radiologist core hydra.searchpath=[file:///abs/path/to/myconfigs] module/loss=my_loss
```

The loss is instantiated **fully** (no `_partial_`) and called as
`loss(logits, target)`.

### A custom metric

```yaml
# myconfigs/module/metric/my_metric.yaml
# @package module
metric:
  _target_: torchmetrics.classification.MulticlassRecall
  _partial_: true
  num_classes: ${module.net.num_classes}
  average: macro
  compute_on_cpu: true
  sync_on_compute: true
```

The metric config **must** carry `_partial_: true`: `LModule` calls the result
twice with no arguments to build separate validation and test metric instances.
The returned object must be a `torchmetrics.Metric` — `LModule` reads
`compute_on_cpu`, `sync_on_compute` and `process_group` off it to configure the
loss trackers consistently.

It is updated as `metric(preds, target)` where `preds` are **hard class indices**
(`logits.argmax(dim=1)`), so choose a label-based metric. Probability-based
metrics (AUROC, calibration error) will compose fine at config time and produce
meaningless numbers at run time.

### A custom backbone

The top-level `module` group member owns the whole `LModule` config, so a new
backbone is a new file at the group root:

```yaml
# myconfigs/module/my_net.yaml
# @package module
_target_: radiologist.core.LModule
net:
  _target_: my_package.nets.MyBackbone
  num_classes: 3
  pretrained: true
trainable_layers:
  "": null            # dot-path into `net`; null unfreezes the whole submodule
priors: null
```

```bash
radiologist core hydra.searchpath=[file:///abs/path/to/myconfigs] module=my_net
```

`net` may be any `nn.Module` returning logits of shape `(B, num_classes)`. Two
things to keep in sync when you swap it:

- `module/metric/fbeta_score.yaml` interpolates `${module.net.num_classes}`, so
  your `net` config must expose a `num_classes` key under that exact path.
- `callbacks/attribution.yaml` and `callbacks/onnx_export.yaml` reference a
  `cam_target_layer` / `target_layer` by dot-path (`layer4` for ResNet). Override
  it for your architecture, or drop those callbacks.

### Other groups you can extend the same way

`module/optimizer`, `module/scheduler` (both `_partial_: true`), `callbacks`,
`loggers`, `strategy`, `debug`, and `datamodule` all follow the identical
pattern — add a file to your search-path directory with the matching
`# @package` header and select it by name.

### What is *not* config-swappable

`LModule` itself resolves `net`, `loss`, `metric`, `optimizer` and `scheduler`
from fixed key names, and the training/validation/test step bodies are not
extension points. Changing what happens inside a step means subclassing
`LModule` and pointing `module._target_` at your subclass.
`````

### AC6 — Four small corrections

- **Line 35**: change the layout comment to `` train.py  # train() orchestration function (the Hydra entry point lives in radiologist-cli) ``
- **Line 185**: replace "or the `radiologist-registry push` / `promote` CLI" with "or the `radiologist registry push` / `radiologist registry promote` commands"
- **Line 197**: replace the cell with `` ResNet-50 backbone; `num_classes: 3` is set here and interpolated *by* `module/metric/fbeta_score.yaml` ``
- **Line 202**: replace the cell with `BestMetric, WandbSummary, Attribution, ModelCheckpoint, LRMonitor, OnnxExport`

---

## Verification

Every claim above was checked against the source, and the `LModule` snippet in AC2 / AC4 was **executed**: `training_step` returned a loss and `configure_optimizers()["optimizer"]` was an `AdamW`. The AC5 config-group mechanism was executed end to end — a custom `module=`, `module/loss=` and `module/metric=` from an external directory all composed correctly via `hydra.searchpath`, and omitting the `# @package module` header reproduced exactly the silent misplacement warned about in the callout.

## Definition of done

- [ ] All six ACs applied to `radiologist-core/README.md`
- [ ] No source, config, or CLI file changed
- [ ] `mkdocs build` succeeds (`make docs-build`)
